### ISSUE TITLE
Ardour + Cocoa rect workaround + Monterey 12.6 => crash

### ISSUE THREAD

**Issue by michaelwillis** [AuburnSounds/Dplug#705]
I thought this was just my plugin, but then I was able to reproduce with the CLIP It example plugin. The crash happens upon loading the plugin.

I noticed there is conditional code using the boolean `_drawRectWorkaround` based on the OS version. I'm using MacOS Monterrey 12.6, so the workaround boolean is true in this case.

I ran the debugger and found that it crashes at the following point in the code that interacts with the cocoa library.

In cocoawindow.d:
```
    if (layer)
    {
        // On Big Sur this is technically a no-op, but that reverts the drawRect behaviour!          
        // This workaround is sanctionned by Apple: https://gist.github.com/lukaskubanek/9a61ac71dc0db8bb04db2028f2635779#gistcomment-3901461
        layer.setContentsFormat(kCAContentsFormatRGBA8Uint);
    }
```

This calls into appkit.d, which crashes when it attempts the msgSend here:
```
    void setContentsFormat(NSString fmt)
    {
        alias fun_t = extern(C) void function (id, SEL, id) nothrow @nogc;
        (cast(fun_t)objc_msgSend)(_id, sel!"setContentsFormat:", fmt._id);
    }
```

There are some more stack frames that show assembly code in the debugger, presumably from the cocoa library. The error message is:

```
Exception has occurred.
EXC_BAD_ACCESS (code=1, address=0x14667c30)
```

I tried building against different versions of Dplug, and found that the crash does not happen with Dplug v12.1.13, but does happen with Dplug v12.1.14 and any later version that I tried.

I also tried REAPER on MacOS and Ardour on Linux, both of which work as expected and do not exhibit this bug.

**Comment by p0nce** [comment]
Do they crash other hosts too in macOS 12.6? All plugin formats in all hosts will have the workaround, but maybe the workaround was made illegal by Apple.

**Comment by p0nce** [comment]
> macOS Monterey 12.6. Released September 12, 2022

This is really concerning...

**Comment by p0nce** [comment]
>  REAPER on MacOS

Mmmmmm. Or it is something completely different, as it's a segfault in the first place.

**Comment by michaelwillis** [comment]
> **Likely fix:** remove workaround on Monterey or later, make it Big Sur only. You can already test that in your system, it should without the workaround.

I'll try removing the workaround, hopefully some time today.

> **Alt fix** why is it Ardour-only?

That's what I was wondering! I don't know much about the apple libraries, but I'm wondering if Ardour does something slightly different with the graphics context. There must be some different state that results in the segfault, but I think that removing the workaround may still fix it without causing the slow performance that was observed in Big Sur.

**Comment by p0nce** [comment]
Would also help to know if this is happening in arm64 or x86_64 arch.

**Comment by michaelwillis** [comment]
> Would also help to know if this is happening in arm64 or x86_64 arch.

I observed this problem on x86_64 architecture.

Also, I tried forcing `_drawRectWorkaround` to false, and it fixed the bug. However, it might still cause the performance issue that was fixed by the workaround, I'm not sure how to profile that.

**Comment by p0nce** [comment]
The performance issue would definitely show up in Activity Monitor, especially on Distort, this is a huge redraw of the whole plugin area, all the time.

**Comment by michaelwillis** [comment]
I tested CPU usage of the distort plugin, there seems to only be increased CPU usage when the UI needs to be redrawn, like when adjusting the dials or if the meters are active.

**Comment by p0nce** [comment]
Reproduced on Monterey 12.5.1, Lens crash Ardour, but SPAN and Nova do not.

**Comment by p0nce** [comment]
> ARDOUR_DLL_PATH not set in environment - exiting

Ardour doesn't run in debugger... fun

**Comment by p0nce** [comment]
Some measurements with and without workarounds, and also with no workarounds but "layer-backed"

```
Lens with workaround (baseline) => 7.9% to 8.3% CPU usage
reaper host: 48.1%
memory 102mb at opening
total charge: 3.81% 8.75%

Lens with layer-backed but without workaround => 8% 
host 48% 
memory 105.3mb at opening
total charge: 3.82% sys 8.57% user

Lens without workaround
host 51.1%
memory 97mb at opening
total charge: sys 4.20% user 8.76%
```

So it's close in territory; there is perhaps a 2.5% to 3% slowdown when not using the workaround in Monterey, this is very far from the Big Sur huge gap.

Subjective test with 4 instance of with workaround vs 4 instance of no-workaround: no-workaround feels a tiny bit better...

Full size test?

**Comment by p0nce** [comment]
Full size: no meaning full difference in load or subjectivity.
We can simply disable the workaround on Monterey and forward, which means saving a few megabytes per plugin, and only have a problem with Ardour + Big Sur, a very unlikely combination.

**Comment by p0nce** [comment]
Is fixed as per your suggestion in v12.7.20 8b9ec285ccd6c0e9d745ccc6affe76a77d312b62, you can simply `dub upgrade`.

**Comment by michaelwillis** [comment]
Thanks @p0nce!
